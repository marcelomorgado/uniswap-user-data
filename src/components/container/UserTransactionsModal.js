import React from "react";
import PropTypes from "prop-types";
import TransactionsModal from "../presentational/TransactionsModal";
import { Query } from "react-apollo";
import { gql } from "apollo-boost";

const GET_USER_TRANSACTIONS = gql`
  query Transaction($userId: String!, $cursor: String) {
    transactions(first: 15, where: { user: $userId }, after: $cursor) {
      id
      tx
      event
      block
      timestamp
      exchangeAddress
      tokenAddress
      tokenSymbol
      user
      ethAmount
      tokenAmount
      fee
    }
  }
`;

class UserTransactionsModal extends React.PureComponent {
  state = {
    hasNextPage: true,
    isNextPageLoading: false,
  };

  render() {
    const { userId, open, handleClose } = this.props;
    return (
      <Query query={GET_USER_TRANSACTIONS} variables={{ userId }}>
        {({ loading, error, data, fetchMore }) => {
          if (loading) return <p>Loading...</p>;
          if (error) return <p>Error :(</p>;
          const { transactions } = data;

          const onLoadMore = () => {
            const lastTransaction = transactions[transactions.length - 1];
            const { id: cursor } = lastTransaction;

            this.setState({ isNextPageLoading: true });

            return fetchMore({
              variables: {
                cursor,
              },
              updateQuery: (prevResult, { fetchMoreResult: newData }) => {
                const { transactions: prevTransactions } = prevResult;
                const { transactions: newTransactions } = newData;

                let data;
                if (!newData || !newTransactions.length) {
                  data = prevResult;
                  this.setState({ hasNextPage: false });
                } else {
                  const transactions = [
                    ...prevTransactions,
                    ...newTransactions,
                  ];

                  data = { transactions };
                }

                this.setState({ isNextPageLoading: false });
                return data;
              },
            });
          };

          const hasNextPage = true;
          const isNextPageLoading = false;

          return (
            <TransactionsModal
              open={open}
              handleClose={handleClose}
              userId={userId}
              transactions={transactions}
              onLoadMore={onLoadMore}
              hasNextPage={hasNextPage}
              isNextPageLoading={isNextPageLoading}
            />
          );
        }}
      </Query>
    );
  }
}

UserTransactionsModal.propTypes = {
  open: PropTypes.bool.isRequired,
  handleOpen: PropTypes.func.isRequired,
  handleClose: PropTypes.func.isRequired,
  userId: PropTypes.string,
};

export default UserTransactionsModal;
