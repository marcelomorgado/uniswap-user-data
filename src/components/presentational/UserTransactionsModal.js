import React from "react";
import PropTypes from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import Modal from "@material-ui/core/Modal";
import TransactionsList from "./TransactionsList";

function getModalStyle() {
  const top = 50;
  const left = 50;

  return {
    top: `${top}%`,
    left: `${left}%`,
    transform: `translate(-${top}%, -${left}%)`,
  };
}

const modalStyles = theme => ({
  paper: {
    position: "absolute",
    width: "80%",
    backgroundColor: theme.palette.background.paper,
    boxShadow: theme.shadows[5],
    padding: theme.spacing.unit * 4,
    outline: "none",
  },
});

class UserTransactionsModal extends React.Component {
  render() {
    const {
      classes,
      onLoadMore,
      handleClose,
      open,
      etherTransactions,
      hasNextPage,
      isNextPageLoading,
    } = this.props;

    return (
      <div>
        <Modal
          aria-labelledby="simple-modal-title"
          aria-describedby="simple-modal-description"
          open={open}
          onClose={handleClose}
        >
          <div style={getModalStyle()} className={classes.paper}>
            <TransactionsList
              items={etherTransactions}
              loadNextPage={onLoadMore}
              hasNextPage={hasNextPage}
              isNextPageLoading={isNextPageLoading}
            />
          </div>
        </Modal>
      </div>
    );
  }
}

UserTransactionsModal.propTypes = {
  classes: PropTypes.object.isRequired,
  open: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  onLoadMore: PropTypes.func.isRequired,
  userId: PropTypes.string,
  etherTransactions: PropTypes.array,
  hasNextPage: PropTypes.bool.isRequired,
  isNextPageLoading: PropTypes.bool.isRequired,
};

export default withStyles(modalStyles)(UserTransactionsModal);
