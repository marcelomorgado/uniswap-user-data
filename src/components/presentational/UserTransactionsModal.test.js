import React from "react";
import { shallow } from "enzyme";
import UserTransactionsModal from "./UserTransactionsModal";

it("renders the component", () => {
  const props = {
    open: true,
    handleClose: () => {},
    onLoadMore: () => {},
    userId: "0x123User",
    etherTransactions: [],
    hasNextPage: false,
    isNextPageLoading: false,
  };

  expect(shallow(<UserTransactionsModal {...props} />)).toMatchSnapshot();
});
