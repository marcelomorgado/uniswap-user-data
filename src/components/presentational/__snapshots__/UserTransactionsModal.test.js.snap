// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div>
  <WithStyles(Modal)
    onClose={[Function]}
    open={true}
  >
    <TransactionsList
      hasNextPage={false}
      isNextPageLoading={false}
      items={Array []}
      loadNextPage={[Function]}
    />
  </WithStyles(Modal)>
</div>
`;
